### PR TITLE
Extends Connection Template support to FCNetwork, FCoENetwork and NetworkSet

### LINKED ISSUE #218: connection_template resource does not support NetworkSet, FCoENetwork and FCNetwork
### Description
I'm trying to configure Bandwidth parameter within the NetworkSet and FcNetwork. connection_template resource obviously supports EthernetNetwork only but connection_template can be configured at the three locations: `Ethernet Networks`, `FC Networks` and `Network Sets`

### Environment Details
 - **oneview Cookbook Version:** 2.1.0
 - **OneView Version:** 3.0 Synergy edition

### Intent / Proposal
I suppose that additional methods would do the job. Something like:
```ruby
# oneview/libraries/resource_providers/api200/connection_template_provider.rb

      def load_resource_from_ethernet
        return unless @context.associated_ethernet_network
        @item.data.delete('name')
        ethernet = resource_named(:EthernetNetwork).find_by(@item.client, name: @context.associated_ethernet_network).first
        @item['uri'] = ethernet['connectionTemplateUri']
      end
```

and

```ruby
# oneview/libraries/resource_providers/api200/connection_template_provider.rb

      def create_or_update
        load_resource_from_ethernet
        load_resource_from_network_set
        super
      end
```

and

```ruby
# oneview/resources/connection_template.rb

OneviewCookbook::ResourceBaseProperties.load(self)

default_action :update

property :associated_ethernet_network, String
property :associated_network_set, String

action :update do
  OneviewCookbook::Helper.do_resource_action(self, :ConnectionTemplate, :create_or_update)
end

action :reset do
  OneviewCookbook::Helper.do_resource_action(self, :ConnectionTemplate, :reset)
end
```

and something similar for FC Networks. 

Do you think it's reasonable enhancement?

### PR DESCRIPTION
**Pull request by vranystepan**
### Description
As described in the issue #218 , `ConnectionTemplateProvider` now supports `EthernetNetwork`, `FcNetwork`, `NetworkSet` and `FCoENetwork` 
Rubocop seems to be happy, manipulation with resources works fine - tested with OneView Synergy edition.

Fixes #218

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass (`$ rake test`).
- [x] New functionality has been documented in the README if applicable.
  - [x] New functionality has been thoroughly documented in [examples](https://github.com/HewlettPackard/oneview-chef/tree/master/examples/) (please include helpful comments).
- [x] Changes documented in the [CHANGELOG](https://github.com/HewlettPackard/oneview-chef/blob/master/CHANGELOG.md).

### CODE REVIEW

File: libraries/resource_providers/api200/connection_template_provider.rb
Comment (fgbulsoni) [inline libraries/resource_providers/api200/connection_template_provider.rb]: We should probably join lines 35-36:
@item['uri'] = load_resource(resource_class_name, resource_name, :connectionTemplateUri)

File: libraries/resource_providers/api200/connection_template_provider.rb
Comment (fgbulsoni) [inline libraries/resource_providers/api200/connection_template_provider.rb]: This `if` is kind of big, I'd suggest one of two things for improvement:
1. Turn this method more generic and just call it, as in:
```ruby
def load_connection_template(resource_type, resource_name)
  return unless resource_name
  @item.data.delete('name')
  @item['uri'] = load_resource(resource_type, resource_name, :connectionTemplateUri)
end
```
then we'd just call `load_connection_template(:EthernetNetwork, @context.associated_ethernet_network)` , then fc, then fcoe, then net instead of calling the method once.

Or, 

2. maybe use a `case` instead of `if`.

I personally kind of like the option 1 better.

File: libraries/resource_providers/api200/connection_template_provider.rb
Comment (fgbulsoni) [inline libraries/resource_providers/api200/connection_template_provider.rb]: Also, important to notice it should be :FCNetwork when calling it

**Review by fgbulsoni — CHANGES_REQUESTED** [review]
A few code changes/improvements pointed out. 
Also this PR should contain unit tests, changelog & readme updates to be mergeable.

**Comment by vranystepan** [comment]
Thank you. Will adjust.

File: libraries/resource_providers/api200/connection_template_provider.rb
Comment (vranystepan) [reply libraries/resource_providers/api200/connection_template_provider.rb]: Oh, I see ... That's actually brilliant idea.

**Comment by vranystepan** [comment]
@fgbulsoni I've made some changes. I've created extra method `load_connection_templates` where I'm calling `load_template_from_resource`. I've done so because we're loading these templates in two actions: `create_or_update` and `reset`. Tests are passing.
Tomorrow morning **I'm gonna test them against the Synergy**.
Will send an update once it's tested.

**Comment by vranystepan** [comment]
# Functionality tests:
## FC networks
**code**
```ruby
oneview_fc_network 'Fibre Channel A' do
  client oneview_sdk_client
  data(
    fabricType: 'FabricAttach',
    autoLoginRedistribution: true
  )
end

oneview_connection_template 'Update connection template: Fibre Channel A' do
   client oneview_sdk_client
   associated_fc_network 'Fibre Channel A'
   data(
     bandwidth: {
       maximumBandwidth: 1000 * 8,
       typicalBandwidth: 1000 * 8
     }
   )
end
```

**Result:**
fc

## Network sets
**code:**
```ruby
oneview_network_set 'network_set' do
  client oneview_sdk_client
  ethernet_network_list ['network_1', 'network_2']
end

oneview_connection_template 'Update connection template: network_set' do
   client oneview_sdk_client
   associated_network_set 'network_set'
   data(
     bandwidth: {
       maximumBandwidth: 1000 * 10,
       typicalBandwidth: 1000 * 10
     }
   )
end

```

**Result:**
ns

overall, I'm happy with the functionality :)

File: CHANGELOG.md
Comment (jsmartt) [inline CHANGELOG.md]: Please attach the issue in this bullet. See below for examples.

This PR also needs some unit tests. If you need some help with those, we can assist. It's exactly not a trivial test setup.

Comment (vranystepan) [reply CHANGELOG.md]: @jsmartt Thanks, I'll definitely need some help with unit tests. What exactly needs to be done? Is it something in here? https://github.com/HewlettPackard/oneview-chef/tree/master/spec I'm sorry for these questions :)

Comment (fgbulsoni) [reply CHANGELOG.md]: Hey guys, I've completed the unit tests and will be pushing to this PR soon. Though, it's good to note for future reference that what we usually mean by `New functionality includes testing.` is actually the inclusion of those unit tests referring to what you might have implemented or changed. Perhaps we might make that more explicit in the template.

File: libraries/resource_providers/api200/connection_template_provider.rb
Comment (fgbulsoni) [inline libraries/resource_providers/api200/connection_template_provider.rb]: Verifying on line 25-26 if only one associated resource has been passed in, if not raising an error. 
This is supposed to prevent the unexpected behavior of just the last associated resource specified being updated instead of all or none.
This way, we enforce what is being said in the readme and the chef way: this operation is for a single resource.

File: .rubocop.yml
Comment (fgbulsoni) [inline .rubocop.yml:41]: Rubocop actually forced reduce operations to use var names 'a' and 'e' in a block, as in: 
```ruby
(...).reduce(0) { |a, e| (...)
```
This rule has been invalidated in the ruby style guide and also disabled in future versions of rubocop, so disabling this here for improved readability.

**Comment by fgbulsoni** [comment]
@jsmartt Please review again since I made some changes to the PR. Once we have another approval I'll also give mine.

File: spec/unit/resources/connection_template/update_spec.rb
Comment (jsmartt) [inline spec/unit/resources/connection_template/update_spec.rb]: How are we validating that it's updating for each resource? Do we need to expect it to receive it a certain number of times? (4) Something like:

```ruby
update_count = 0
allow_any_instance_of(OneviewSDK::ConnectionTemplate).to receive(:update) { |_i| update_count +=1 }
...
# Then on the last line of the test:
expect(update_count).to eq(4)
```

Comment (fgbulsoni) [reply spec/unit/resources/connection_template/update_spec.rb]: Awesome idea, I was bothered by the allow as well but couldn't fit an expect there due to multiple resource instances.
Updating the PR atm.

**Comment by fgbulsoni** [comment]
@jsmartt There you go, updated according to comments. Please review again whenever you have the time.
:octocat: :taco:

File: libraries/resource_providers/api200/connection_template_provider.rb
Comment (tmiotto) [inline libraries/resource_providers/api200/connection_template_provider.rb]: It will be more readable if  resources_set is simply the number of resources set, and you use `if resources_set > 1`

Comment (jsmartt) [reply libraries/resource_providers/api200/connection_template_provider.rb]: lol, I had to do a bit of digging into this to understand it too. It works, but maybe something like this is more readable...

```ruby
resources_set = [@context.associated_ethernet_network, @context.associated_fcoe_network, @context.associated_fc_network,
                 @context.associated_network_set].compact.size
raise 'A single associated resource field must be specified for this action.' unless resources_set == 1
```

Update: Or maybe it should be `> 1`. Depends on if it can be 0.

Comment (fgbulsoni) [reply libraries/resource_providers/api200/connection_template_provider.rb]: It should be `> 1` since we can just update the connection template directly without getting it from an associated resource.
Liked the solution though, didn't know about `compact`. @tmiotto and I had to bump heads a little to come up with the map&reduce for this.

**Comment by fgbulsoni** [comment]
@jsmartt @tmiotto Hopefully now we've covered all fronts in this PR. Please let me know if there's anything else that catches your attention or if this is good to go. :octocat:

**Review by jsmartt — APPROVED** [review]
Looks good to me 👍

**Review by fgbulsoni — APPROVED** [review]

File: CHANGELOG.md
Comment (tmiotto) [inline CHANGELOG.md]: FCNetwork

**Review by tmiotto — CHANGES_REQUESTED** [review]
Just change it and it will be good to go

**Review by tmiotto — APPROVED** [review]
:taco: :shipit:

**Comment by jsmartt** [comment]
Is this ready to be merged?

**Comment by fgbulsoni** [comment]
> Is this ready to be merged?

It is, I was just waiting for Travis to finish. Merging atm the moment.

**Comment by vranystepan** [comment]
Thank you very much! You're my superheroes :rocket:

**Pull request merged**
